src/kvstore/NebulaStore.cpp

```diff
@@ -118,43 +112,93 @@ void NebulaStore::loadPartFromDataPath() {
             enginePtr->removePart(partId);
             continue;
           } else {
             auto spacePart = std::make_pair(spaceId, partId);
             if (spacePartIdSet.find(spacePart) == spacePartIdSet.end()) {
               spacePartIdSet.emplace(spacePart);
-              partIds.emplace_back(partId);
+              partPeers.emplace(partId, peers);
             }
           }
         }
-        if (partIds.empty()) {
+        if (partPeers.empty()) {
+          if (!options_.partMan_->spaceExist(storeSvcAddr_, spaceId).ok()) {
+            if (FLAGS_auto_remove_invalid_space) {
+              auto spaceDir = folly::stringPrintf("%s/%s", rootPath.c_str(), dir.c_str());
+              removeSpaceDir(spaceDir);
+            }
+          }
           continue;
         }
 
-        std::atomic<size_t> counter(partIds.size());
+        std::atomic<size_t> counter(partPeers.size());
         folly::Baton<true, std::atomic> baton;
-        LOG(INFO) << "Need to open " << partIds.size() << " parts of space " << spaceId;
-        for (auto& partId : partIds) {
-          bgWorkers_->addTask([spaceId, partId, enginePtr, &counter, &baton, this]() mutable {
-            auto part = newPart(spaceId, partId, enginePtr, false, {});
-            LOG(INFO) << "Load part " << spaceId << ", " << partId << " from disk";
-
-            {
-              folly::RWSpinLock::WriteHolder holder(&lock_);
-              auto iter = spaces_.find(spaceId);
-              CHECK(iter != spaces_.end());
-              iter->second->parts_.emplace(partId, part);
-            }
-            counter.fetch_sub(1);
-            if (counter.load() == 0) {
-              baton.post();
-            }
-          });
+        LOG(INFO) << "Need to open " << partPeers.size() << " parts of space " << spaceId;
+        for (auto& it : partPeers) {
+          auto& partId = it.first;
+          Peers& peers = it.second;
+
+          bgWorkers_->addTask(
+              [spaceId, partId, &peers, enginePtr, &counter, &baton, this]() mutable {
+                // create part
+                bool isLearner = false;
+                std::vector<HostAddr> addrs;  // raft peers
+                for (auto& [addr, peer] : peers.getPeers()) {
+                  if (addr == raftAddr_) {  // self
+                    if (peer.status == Peer::Status::kLearner) {
+                      isLearner = true;
+                    }
+                  } else {  // others
+                    if (peer.status == Peer::Status::kNormalPeer ||
+                        peer.status == Peer::Status::kPromotedPeer) {
+                      addrs.emplace_back(addr);
+                    }
+                  }
+                }
+                auto part = newPart(spaceId, partId, enginePtr, isLearner, addrs);
+                LOG(INFO) << "Load part " << spaceId << ", " << partId << " from disk";
+
+                // add learner peers
+                if (!isLearner) {
```

Comment:
When the current machine is a learner machine, is the learner part not added to the raft group?